### DISCUSSION TITLE
JSON column sorting

### DISCUSSION THREAD

**Discussion by keizah7** [laravel/nova-issues#5105] [Q&A]
- Laravel Version: 9.37
- Nova Version: 4.17
- PHP Version: 8.1
- Database Driver & Version: MySQL 8
- Operating System and Version: Ventura 13
- Browser type and version: Firefox 107

### Description:

`sortable()` isn't working with [spatie/laravel-translatable](https://github.com/spatie/laravel-translatable) column:

```php
class OrderItem extends Model
{
    use HasFactory;
    use HasTranslations;

    protected $guarded = ['id'];

    public array $translatable = ['title'];
```
```php
Text::make(__('Title'), 'title')
    ->rules('required', 'max:255')
    ->sortable(),
```

```php
$table->json('title');
```

```json
{
  "lt": "A",
  "en": "X"
}
```

![image](https://user-images.githubusercontent.com/49227202/203087675-a2325e35-ffe0-45fa-a404-73c719254dff.png)
Column has sorting in UI, query is changing from asc to desc. but it doesn't sort the data.

Is this a bug? Because I can't pass callback to `sortable` method for manual order resolving, eg:

```php
->sortable(function ($query) {
    $locale = app()->getLocale();

    return $query->orderBy('attribute_title->' . $locale);
})
```

**Comment by crynobone** [comment]
1. `sortable()` doesn't support callable and if it does it would only be `@param \Closure(\Laravel\Nova\Http\Requests\NovaRequest):bool $callback`.
2. Laravel Nova uses `Field::sortableUriKey()` and that would be use in the `orderBy` query.

**Comment by nrueckmann** [comment]
We just need the ability to pass the `sortableUriKey`.

Example (Field.):

```
<?php

namespace Laravel\Nova\Fields;

 ...

abstract class Field extends FieldElement implements JsonSerializable, Resolvable
{
    ...

    private $sortableUriKey = null;

    /**
     * Specify that this field should be sortable.
     *
     * @param  bool  $value
     * @return $this
     */
    public function sortable($value = true, $sortableUriKey = null)
    {
        if (! $this->computed()) {
            $this->sortable = $value;
        }

        if ($sortableUriKey) {
            $this->sortableUriKey = $sortableUriKey;
        }

        return $this;
    }

    /**
     * Return the sortable uri key for the field.
     *
     * @return string
     */
    public function sortableUriKey()
    {
        return $this->sortableUriKey ?? $this->attribute;
    }

    ...
}
```
